src/lifecycle.rs

```diff
@@ -0,0 +1,50 @@
+use std::mem;
+
+use neon_runtime::raw::Env;
+use neon_runtime::reference;
+use neon_runtime::tsfn::ThreadsafeFunction;
+
+use crate::context::Context;
+use crate::handle::root::NapiRef;
+
+pub(crate) struct InstanceData {
+    drop_queue: &'static ThreadsafeFunction<NapiRef>,
+}
+
+fn drop_napi_ref(env: Option<Env>, data: NapiRef) {
+    if let Some(env) = env {
+        unsafe {
+            reference::unreference(env, mem::transmute(data));
+        }
+    }
+}
+
+impl InstanceData {
+    pub(crate) fn get<'a, C: Context<'a>>(cx: &mut C) -> &'a mut InstanceData {
+        let env = cx.env().to_raw();
+        let data =
+            unsafe { neon_runtime::lifecycle::get_instance_data::<InstanceData>(env).as_mut() };
+
+        if let Some(data) = data {
+            return data;
+        }
+
+        let drop_queue = Box::new(unsafe {
+            let mut queue = ThreadsafeFunction::new(env, drop_napi_ref);
+            queue.unref(env);
+            queue
+        });
+
+        let data = InstanceData {
+            drop_queue: Box::leak(drop_queue),
```

Comment:
We talked about this a couple days ago. I think I prefer the alternative approach you mentioned of using an `Arc`, and we can consider optimizing in the future.

Reply:
I will make that change and add some documentation on the design in the form of a code comment. Hopefully it will help hint a possible solution if it _does_ pop up as a bottleneck for a user.